### ISSUE TITLE
Fiona support for vsimem datasets?

### ISSUE THREAD

**Issue by youngpm** [Toblerity/Fiona#211]
I noticed that fiona can handle some instances of virtual file systems, e.g.

http://toblerity.org/fiona/manual.html#virtual-filesystems

What about datasets constructed via GDAL's vsimem mechanism?  

My use case is that I'm getting a response back from an ArcServer endpoint and the geometry is ESRI flavored JSON.  In OGR, I can do something like 
```python
r = requests.get(arc_server_url, params=my_json_request)
vsi_filename = '/vsimem/my_virtual_file.json'
gdal.FileFromMemBuffer(vsi_filename, r.content)
ds = ogr.Open(vsi_filename)
lyr = ds.GetLayer()
geojson_feats = [json.loads(feat.ExportToJson()) for feat in lyr]
ds = None 
gdal.Unlink(vsi_filename)
```
This is a bit clunky IMO, but maybe Fiona can do better since it already supports zipped virtual file systems?

**Comment by sgillies** [comment]
Let see if I got this right: given a byte string that contains JSON, you'd like to open it in Fiona as if it were a file and read out GeoJSON-style features?

There's already precedent in Python for this kind of thing: [stringio](https://docs.python.org/2/library/stringio.html). Perhaps the thing to do for Fiona would be to write a read-only StringIO-like class extending `fiona.collection.Collection` which uses vsimem internally. Something like

```
resp = requests.get(arc_server_url, params=my_json_request)
for feature in fiona.collection.BytesCollection(resp.content):
    print feature['id'], feature['properties'], feature['geometry']
```

should be quite possible. Would you be interested in helping develop and test such a thing?

**Comment by youngpm** [comment]
Yeah!  I think such functionality would be useful beyond getting these ERSI flavored JSON responses cracked open; basically if you get some buffer back from a response that's in a format that OGR can handle, it would be extremely satisfying to read it with Fiona!

This is a good chance for me to pick up some Cython too; I will start poking see how far I get!

**Comment by sgillies** [comment]
Let me know how it goes. Submitting a PR early on will help me and the other Fiona track your work and give you crucial feedback.

We don't have anything in the wiki yet, but here's a quick outline of dev practice:

- We're targeting GDAL 1.10-11. No 2.0 yet.
- GDAL/OGR functions we call are spec'd in https://github.com/Toblerity/Fiona/blob/master/fiona/ograpi.pxd. It's like our single header file.
- Cython extension modules are named like "_foo.pyx"
- Python extension modules named foo.py then import from _foo. This separation (hopefully) allows us some bug and docstring fixes from users who don't know or don't have time for Cython.
- We use nose for testing (I actually use pytest, which is almost entirely compatible, but Travis runs nosetests).

This will be a cool feature. I'm curious about doing something similar for rasters.

-

**Comment by youngpm** [comment]
I'm starting to make some progress.  I've wrapped the handful of functions I need out out cpl_vsi.h;  I must admit I find the GDAL docs on the VSI system to be somewhat lacking!  I did come across this interesting tidbit in item 5 on [GDAL's 2015 Summer of Code](http://trac.osgeo.org/gdal/wiki/SummerOfCode).  If the VSI stuff is moved to an external project, it could be wrapped up as a separate project and used by both Fiona and Rasterio.

Anyways, my current thinking is to subclass [Session](https://github.com/Toblerity/Fiona/blob/master/fiona/ogrext.pyx#L324) such that it will be responsible for cleaning up the created in-mem file, and then a subclass of [Collection](https://github.com/Toblerity/Fiona/blob/master/fiona/collection.py#L15) will use that Session?  I'm still exploring the code so we'll see what shakes out; when I have a rough prototype I'll do a PR for feedback.
